File: greenlang/factors/sdk/ts/src/transport.ts
/**
 * HTTP transport layer for the Factors SDK.
 *
 * Parity with `greenlang/factors/sdk/python/transport.py`:
 *
 *   - Exponential backoff on 429 / 5xx / network errors
 *   - Retry-After header honoring
 *   - Transparent ETag response cache (If-None-Match)
 *   - X-Request-ID / X-RateLimit-* / X-Factors-Edition parsing
 *   - Error mapping via `errorFromResponse`
 *
 * The transport is built on the native `fetch` API which is available
 * in modern browsers and Node.js 18+. For older Node releases, users
 * can install `undici` and pass a polyfilled `fetch` via the
 * `fetchImpl` option.
 */

import {
  AuthContext,
  AuthProvider,
  composeAuthHeaders,
} from './auth';
import { canonicalJsonCompactBytes } from './canonical';
import { FactorsAPIError, errorFromResponse } from './errors';

export const DEFAULT_TIMEOUT_MS = 30_000;
export const DEFAULT_MAX_RETRIES = 3;
export const DEFAULT_USER_AGENT = 'greenlang-factors-sdk-ts/1.0.0';

// -----------------------------------------------------------------------------
// Fetch typing (lets us accept custom fetch impls like undici's without a
// hard dependency on DOM lib types).
// -----------------------------------------------------------------------------

export type FetchLike = (
  input: string,
  init?: {
    method?: string;
    headers?: Record<string, string>;
    body?: Uint8Array | string;
    signal?: AbortSignal;
  },
) => Promise<FetchResponseLike>;

export interface FetchResponseLike {
  status: number;
  statusText: string;
  ok: boolean;
  headers: { get(name: string): string | null };
  text(): Promise<string>;
  url?: string;
}

function getGlobalFetch(): FetchLike {
  const g = globalThis as unknown as { fetch?: FetchLike };
  if (typeof g.fetch !== 'function') {
    throw new Error(
      'No global fetch() available. Pass a FetchLike via `fetchImpl` ' +
        '(e.g. `(await import("undici")).fetch`) or run on Node 18+ / a browser.',
    );
  }
  return g.fetch.bind(globalThis) as FetchLike;
}

// -----------------------------------------------------------------------------
// Rate-limit info
// -----------------------------------------------------------------------------

export interface RateLimitInfo {
  limit?: number;
  remaining?: number;
  reset?: number;
  retryAfter?: number;
}

function parseRateLimit(headers: { get(k: string): string | null }): RateLimitInfo {
  const toInt = (v: string | null): number | undefined => {
    if (v === null) return undefined;
    const n = parseInt(v, 10);
    return Number.isFinite(n) ? n : undefined;
  };
  const toFloat = (v: string | null): number | undefined => {
    if (v === null) return undefined;
    const n = parseFloat(v);
    return Number.isFinite(n) ? n : undefined;
  };
  return {
    limit: toInt(headers.get('X-RateLimit-Limit')),
    remaining: toInt(headers.get('X-RateLimit-Remaining')),
    reset: toInt(headers.get('X-RateLimit-Reset')),
    retryAfter: toFloat(headers.get('Retry-After')),
  };
}

// -----------------------------------------------------------------------------
// ETag cache
// -----------------------------------------------------------------------------

interface ETagEntry {
  etag: string;
  data: unknown;
  headers: Record<string, string>;
}

export class ETagCache {
  private readonly entries = new Map<string, ETagEntry>();
  private readonly maxEntries: number;

  constructor(maxEntries: number = 512) {
    this.maxEntries = Math.max(1, maxEntries);
  }

  static key(
    method: string,
    url: string,
    params?: Record<string, unknown> | null,
  ): string {
    let qs = '';
    if (params) {
      const pairs: Array<[string, string]> = [];
      for (const [k, v] of Object.entries(params)) {
        if (v === undefined || v === null) continue;
        pairs.push([k, String(v)]);
      }
      pairs.sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0));
      qs = pairs.map(([k, v]) => `${k}=${v}`).join('&');
    }
    return `${method.toUpperCase()} ${url}?${qs}`;
  }

  get(key: string): ETagEntry | undefined {
    return this.entries.get(key);
  }

  set(
    key: string,
    etag: string,
    data: unknown,
    headers: Record<string, string> = {},
  ): void {
    if (this.entries.size >= this.maxEntries) {
      const firstKey = this.entries.keys().next().value;
      if (firstKey !== undefined) this.entries.delete(firstKey);
    }
    this.entries.set(key, { etag, data, headers: { ...headers } });
  }

  clear(): void {
    this.entries.clear();
  }

  size(): number {
    return this.entries.size;
  }
}

// -----------------------------------------------------------------------------
// Transport response
// -----------------------------------------------------------------------------

export interface TransportResponse<T = unknown> {
  statusCode: number;
  data: T;
  headers: Record<string, string>;
  etag?: string;
  fromCache: boolean;
  requestId?: string;
  rateLimit: RateLimitInfo;
  edition?: string;
}

// -----------------------------------------------------------------------------
// Transport options & class
// -----------------------------------------------------------------------------

export interface TransportOptions {
  baseUrl: string;
  auth?: AuthProvider;
  timeoutMs?: number;
  maxRetries?: number;
  userAgent?: string;
  defaultEdition?: string;
  cache?: ETagCache;
  extraHeaders?: Record<string, string>;
  fetchImpl?: FetchLike;
  /** Override backoff (ms) for tests. */
  sleep?: (ms: number) => Promise<void>;
}

function defaultSleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function shouldRetry(status: number): boolean {
  return status === 429 || (status >= 500 && status < 600);
}

function computeWait(attempt: number, retryAfterSec?: number): number {
  if (retryAfterSec !== undefined && retryAfterSec >= 0) {
    return Math.min(retryAfterSec, 60) * 1000;
  }
  return Math.min(2 ** (attempt - 1), 30) * 1000;
}

function decodeBody(text: string, contentType: string): unknown {
  const ct = (contentType || '').toLowerCase();
  const trimmed = text.trimStart();
  if (
    ct.includes('application/json') ||
    trimmed.startsWith('{') ||
    trimmed.startsWith('[')
  ) {
    try {
      return JSON.parse(text);
    } catch {
      return text;
    }
  }
  return text;
}

function headersToRecord(
  headers: { get(name: string): string | null },
  knownNames: readonly string[],
): Record<string, string> {
  const out: Record<string, string> = {};
  for (const n of knownNames) {
    const v = headers.get(n);
    if (v !== null) out[n] = v;
  }
  return out;
}

const KNOWN_HEADER_NAMES = [
  'ETag',
  'Content-Type',
  'Content-Length',
  'X-Request-ID',
  'X-RateLimit-Limit',
  'X-RateLimit-Remaining',
  'X-RateLimit-Reset',
  'Retry-After',
  'X-Factors-Edition',
  'X-GreenLang-Edition',
  'Cache-Control',
] as const;

export class Transport {
  public readonly baseUrl: string;
  public readonly cache: ETagCache;

  private readonly auth?: AuthProvider;
  private readonly timeoutMs: number;
  private readonly maxRetries: number;
  private readonly userAgent: string;
  private readonly defaultEdition?: string;
  private readonly extraHeaders: Record<string, string>;
  private readonly fetchImpl: FetchLike;
  private readonly sleep: (ms: number) => Promise<void>;

  constructor(opts: TransportOptions) {
    if (!opts.baseUrl) throw new Error('Transport requires a baseUrl');
    this.baseUrl = opts.baseUrl.replace(/\/+$/, '');
    this.auth = opts.auth;
    this.timeoutMs = opts.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    this.maxRetries = Math.max(1, opts.maxRetries ?? DEFAULT_MAX_RETRIES);
    this.userAgent = opts.userAgent ?? DEFAULT_USER_AGENT;
    this.defaultEdition = opts.defaultEdition;
    this.cache = opts.cache ?? new ETagCache();
    this.extraHeaders = { ...(opts.extraHeaders ?? {}) };
    this.fetchImpl = opts.fetchImpl ?? getGlobalFetch();
    this.sleep = opts.sleep ?? defaultSleep;
  }

  private buildUrl(path: string, params?: Record<string, unknown> | null): string {
    const norm = path.startsWith('/') ? path : '/' + path;
    let url = this.baseUrl + norm;
    if (params) {
      const entries = Object.entries(params).filter(
        ([, v]) => v !== undefined && v !== null,
      );
      if (entries.length > 0) {
        const qs = entries
          .flatMap(([k, v]) => {
            if (Array.isArray(v)) {
              return v.map(
                (vv) =>
                  encodeURIComponent(k) + '=' + encodeURIComponent(String(vv)),
              );
            }
            return [
              encodeURIComponent(k) + '=' + encodeURIComponent(String(v)),
            ];
          })
          .join('&');
        url += '?' + qs;
      }
    }
    return url;
  }

  private async headersFor(
    method: string,
    path: string,
    body: Uint8Array | undefined,
    extra?: Record<string, string>,
  ): Promise<Record<string, string>> {
    const base: Record<string, string> = {
      Accept: 'application/json',
      'User-Agent': this.userAgent,
      ...this.extraHeaders,
    };
    if (this.defaultEdition) base['X-Factors-Edition'] = this.defaultEdition;
    if (extra) Object.assign(base, extra);
    const ctx: AuthContext = {
      method: method.toUpperCase(),
      path: path.startsWith('/') ? path : '/' + path,
      body,
    };
    return composeAuthHeaders(this.auth, base, ctx);
  }

  async request<T = unknown>(
    method: string,
    path: string,
    opts: {
      params?: Record<string, unknown> | null;
      jsonBody?: unknown;
      extraHeaders?: Record<string, string>;
      useCache?: boolean;
    } = {},
  ): Promise<TransportResponse<T>> {
    const useCache = opts.useCache ?? true;
    const normPath = path.startsWith('/') ? path : '/' + path;
    const url = this.buildUrl(normPath, opts.params ?? null);
    const cacheKey = ETagCache.key(method, normPath, opts.params ?? null);

    let bodyBytes: Uint8Array | undefined;
    if (opts.jsonBody !== undefined) {
      // Compact canonical JSON matches Python transport byte-for-byte so
      // HMAC body digests are identical across clients.
      bodyBytes = canonicalJsonCompactBytes(opts.jsonBody);
    }

    const cached =
      useCache && method.toUpperCase() === 'GET'
        ? this.cache.get(cacheKey)
        : undefined;

    const condExtra: Record<string, string> = { ...(opts.extraHeaders ?? {}) };
    if (cached && !('If-None-Match' in condExtra)) {
      condExtra['If-None-Match'] = cached.etag;
    }

    let headers = await this.headersFor(method, normPath, bodyBytes, condExtra);
    if (bodyBytes !== undefined) {
      headers = { ...headers, 'Content-Type': 'application/json' };
    }

    let lastErr: Error | undefined;
    for (let attempt = 1; attempt <= this.maxRetries; attempt++) {
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), this.timeoutMs);

      let response: FetchResponseLike;
      try {
        response = await this.fetchImpl(url, {
          method: method.toUpperCase(),
          headers,
          body: bodyBytes,
          signal: controller.signal,
        });
      } catch (exc) {
        clearTimeout(timer);
        lastErr = exc instanceof Error ? exc : new Error(String(exc));
        if (attempt >= this.maxRetries) {
          throw new FactorsAPIError(`Network error: ${lastErr.message}`, {
            context: { attempts: attempt, url: normPath },
          });
        }
        await this.sleep(computeWait(attempt, undefined));
        continue;
      }
      clearTimeout(timer);

      const rateLimit = parseRateLimit(response.headers);
      const requestId = response.headers.get('X-Request-ID') ?? undefined;
      const editionHeader =
        response.headers.get('X-Factors-Edition') ??
        response.headers.get('X-GreenLang-Edition') ??
        undefined;
      const etag = response.headers.get('ETag') ?? undefined;

      if (response.status === 304 && cached) {
        const hdrs = headersToRecord(response.headers, KNOWN_HEADER_NAMES);
        return {
          statusCode: 200,
          data: cached.data as T,
          headers: hdrs,
          etag: cached.etag,
          fromCache: true,
          requestId,
          rateLimit,
          edition: editionHeader,
        };
      }

      if (shouldRetry(response.status) && attempt < this.maxRetries) {
        const wait = computeWait(attempt, rateLimit.retryAfter);
        await this.sleep(wait);
        continue;
      }

      const text = await response.text();
      const body = decodeBody(text, response.headers.get('Content-Type') ?? '');

      if (response.ok) {
        const hdrs = headersToRecord(response.headers, KNOWN_HEADER_NAMES);
        if (etag && method.toUpperCase() === 'GET' && useCache) {
          this.cache.set(cacheKey, etag, body, hdrs);
        }
        return {
          statusCode: response.status,
          data: body as T,
          headers: hdrs,
          etag,
          fromCache: false,
          requestId,
          rateLimit,
          edition: editionHeader,
        };
      }

      throw errorFromResponse({
        statusCode: response.status,
        url: response.url ?? url,
        body,
        requestId,
        retryAfter: rateLimit.retryAfter,
      });
    }

    if (lastErr) {
      throw new FactorsAPIError(
        `Request failed after ${this.maxRetries} attempts: ${lastErr.message}`,
      );
    }
    throw new FactorsAPIError(
      `Request failed after ${this.maxRetries} attempts`,
    );
  }

  close(): void {
    this.cache.clear();
  }
}

File: greenlang/factors/sdk/python/transport.py
# -*- coding: utf-8 -*-
"""HTTP transport layer for the Factors SDK.

Wraps :mod:`httpx` with:

    * Tenacity-based retry on 429 / 5xx / network errors
    * ``Retry-After`` header honoring
    * Transparent ETag response cache (``If-None-Match``)
    * Request-id + rate-limit header parsing
    * Structured error mapping via :func:`greenlang.factors.sdk.python.errors.error_from_response`

The design deliberately separates sync and async transports so callers
can pick a single code path without paying the coroutine cost.
"""
from __future__ import annotations

import json
import logging
import threading
import time
from dataclasses import dataclass, field
from typing import Any, Dict, Mapping, Optional, Tuple

try:
    import httpx
except ImportError:  # pragma: no cover - handled at runtime with clear error
    httpx = None  # type: ignore[assignment]

try:
    from tenacity import (
        RetryError,
        retry,
        retry_if_exception_type,
        stop_after_attempt,
        wait_exponential,
    )
except ImportError:  # pragma: no cover - handled at runtime with clear error
    retry = None  # type: ignore[assignment]

from .auth import AuthProvider, compose_auth_headers
from .errors import FactorsAPIError, RateLimitError, error_from_response

logger = logging.getLogger(__name__)

DEFAULT_TIMEOUT: float = 30.0
DEFAULT_MAX_RETRIES: int = 3
DEFAULT_USER_AGENT: str = "greenlang-factors-sdk-python/1.0.0"


@dataclass
class RateLimitInfo:
    """Rate-limit headers parsed from a response."""

    limit: Optional[int] = None
    remaining: Optional[int] = None
    reset: Optional[int] = None
    retry_after: Optional[float] = None

    @classmethod
    def from_headers(cls, headers: Mapping[str, str]) -> "RateLimitInfo":
        def _int(v: Optional[str]) -> Optional[int]:
            try:
                return int(v) if v is not None else None
            except (TypeError, ValueError):
                return None

        def _float(v: Optional[str]) -> Optional[float]:
            try:
                return float(v) if v is not None else None
            except (TypeError, ValueError):
                return None

        return cls(
            limit=_int(headers.get("X-RateLimit-Limit")),
            remaining=_int(headers.get("X-RateLimit-Remaining")),
            reset=_int(headers.get("X-RateLimit-Reset")),
            retry_after=_float(headers.get("Retry-After")),
        )


@dataclass
class TransportResponse:
    """Decoded response payload + relevant metadata."""

    status_code: int
    data: Any
    headers: Dict[str, str] = field(default_factory=dict)
    etag: Optional[str] = None
    from_cache: bool = False
    request_id: Optional[str] = None
    rate_limit: RateLimitInfo = field(default_factory=RateLimitInfo)
    edition: Optional[str] = None


# ---------------------------------------------------------------------------
# ETag cache
# ---------------------------------------------------------------------------


class ETagCache:
    """Simple thread-safe in-memory ETag/body cache.

    Keyed on ``(method, url, sorted(params))``.  Not LRU (size-bounded
    by ``max_entries``).  Sufficient for SDK read-mostly workloads; if
    a user needs more, they can plug their own via ``cache=`` in the
    client constructor.
    """

    def __init__(self, max_entries: int = 512) -> None:
        self._lock = threading.Lock()
        self._entries: Dict[str, Tuple[str, Any, Dict[str, str]]] = {}
        self._max_entries = max_entries

    @staticmethod
    def key(method: str, url: str, params: Optional[Mapping[str, Any]] = None) -> str:
        if params:
            items = sorted((k, str(v)) for k, v in params.items() if v is not None)
            param_str = "&".join(f"{k}={v}" for k, v in items)
        else:
            param_str = ""
        return f"{method.upper()} {url}?{param_str}"

    def get(self, key: str) -> Optional[Tuple[str, Any, Dict[str, str]]]:
        with self._lock:
            return self._entries.get(key)

    def set(
        self,
        key: str,
        etag: str,
        data: Any,
        headers: Optional[Mapping[str, str]] = None,
    ) -> None:
        with self._lock:
            if len(self._entries) >= self._max_entries:
                # Drop oldest (insertion order preserved in dict).
                self._entries.pop(next(iter(self._entries)))
            self._entries[key] = (etag, data, dict(headers or {}))

    def clear(self) -> None:
        with self._lock:
            self._entries.clear()


# ---------------------------------------------------------------------------
# Shared helpers
# ---------------------------------------------------------------------------


def _check_httpx() -> None:
    if httpx is None:  # pragma: no cover
        raise ImportError(
            "httpx is required for the Factors SDK. "
            "Install with: pip install 'httpx>=0.27'"
        )


def _decode_body(response: "httpx.Response") -> Any:
    ctype = (response.headers.get("content-type") or "").lower()
    try:
        if "application/json" in ctype or response.text.lstrip().startswith(("{", "[")):
            return response.json()
        return response.text
    except (json.JSONDecodeError, ValueError):
        return response.text


def _extract_response_meta(response: "httpx.Response") -> Dict[str, Any]:
    rate = RateLimitInfo.from_headers(response.headers)
    return {
        "etag": response.headers.get("ETag"),
        "request_id": response.headers.get("X-Request-ID"),
        "edition": response.headers.get("X-Factors-Edition")
        or response.headers.get("X-GreenLang-Edition"),
        "rate_limit": rate,
    }


def _should_retry(status: int) -> bool:
    return status == 429 or (500 <= status < 600)


# ---------------------------------------------------------------------------
# Sync transport
# ---------------------------------------------------------------------------


class Transport:
    """Synchronous HTTP transport backed by :class:`httpx.Client`."""

    def __init__(
        self,
        base_url: str,
        *,
        auth: Optional[AuthProvider] = None,
        timeout: float = DEFAULT_TIMEOUT,
        max_retries: int = DEFAULT_MAX_RETRIES,
        user_agent: str = DEFAULT_USER_AGENT,
        default_edition: Optional[str] = None,
        cache: Optional[ETagCache] = None,
        client: Optional[Any] = None,
        transport: Optional[Any] = None,
        extra_headers: Optional[Mapping[str, str]] = None,
    ) -> None:
        _check_httpx()
        self.base_url = base_url.rstrip("/")
        self.auth = auth
        self.timeout = timeout
        self.max_retries = max(1, int(max_retries))
        self.user_agent = user_agent
        self.default_edition = default_edition
        self.cache = cache if cache is not None else ETagCache()
        self._extra_headers = dict(extra_headers or {})
        self._owns_client = client is None
        if client is None:
            self._client = httpx.Client(
                base_url=self.base_url,
                timeout=timeout,
                transport=transport,
            )
        else:
            self._client = client

    def close(self) -> None:
        if self._owns_client:
            self._client.close()

    def __enter__(self) -> "Transport":
        return self

    def __exit__(self, *exc: Any) -> None:
        self.close()

    def _headers(
        self,
        *,
        method: str,
        path: str,
        body: Optional[bytes],
        extra: Optional[Mapping[str, str]] = None,
    ) -> Dict[str, str]:
        base = {
            "Accept": "application/json",
            "User-Agent": self.user_agent,
        }
        base.update(self._extra_headers)
        if self.default_edition:
            base["X-Factors-Edition"] = self.default_edition
        if extra:
            base.update(extra)
        return compose_auth_headers(
            self.auth, base, method=method, path=path, body=body
        )

    def request(
        self,
        method: str,
        path: str,
        *,
        params: Optional[Mapping[str, Any]] = None,
        json_body: Optional[Any] = None,
        extra_headers: Optional[Mapping[str, str]] = None,
        use_cache: bool = True,
    ) -> TransportResponse:
        url_path = path if path.startswith("/") else "/" + path
        cache_key = ETagCache.key(method, url_path, params)

        body_bytes: Optional[bytes] = None
        if json_body is not None:
            body_bytes = json.dumps(json_body, separators=(",", ":"), sort_keys=True).encode(
                "utf-8"
            )

        # ETag pre-flight: attach If-None-Match when we have a cached entry.
        cached = self.cache.get(cache_key) if (use_cache and method.upper() == "GET") else None
        cond_extra: Dict[str, str] = dict(extra_headers or {})
        if cached is not None and "If-None-Match" not in cond_extra:
            cond_extra["If-None-Match"] = cached[0]

        headers = self._headers(
            method=method, path=url_path, body=body_bytes, extra=cond_extra
        )
        if body_bytes is not None:
            headers["Content-Type"] = "application/json"

        last_exc: Optional[BaseException] = None
        for attempt in range(1, self.max_retries + 1):
            try:
                response = self._client.request(
                    method,
                    url_path,
                    params=dict(params) if params else None,
                    content=body_bytes,
                    headers=headers,
                )
            except httpx.RequestError as exc:
                last_exc = exc
                logger.warning(
                    "Network error on %s %s (attempt %d/%d): %s",
                    method, url_path, attempt, self.max_retries, exc,
                )
                if attempt >= self.max_retries:
                    raise FactorsAPIError(
                        "Network error: %s" % exc,
                        context={"attempts": attempt, "url": url_path},
                    ) from exc
                self._sleep_backoff(attempt)
                continue

            meta = _extract_response_meta(response)
            rate: RateLimitInfo = meta["rate_limit"]

            if response.status_code == 304 and cached is not None:
                logger.debug("ETag cache hit (304) for %s", cache_key)
                return TransportResponse(
                    status_code=200,
                    data=cached[1],
                    headers=dict(response.headers),
                    etag=cached[0],
                    from_cache=True,
                    request_id=meta["request_id"],
                    rate_limit=rate,
                    edition=meta["edition"],
                )

            if _should_retry(response.status_code) and attempt < self.max_retries:
                wait = self._compute_wait(attempt, rate.retry_after)
                logger.warning(
                    "Retrying %s %s after HTTP %d in %.2fs (attempt %d/%d)",
                    method, url_path, response.status_code, wait, attempt, self.max_retries,
                )
                time.sleep(wait)
                continue

            body = _decode_body(response)

            if response.is_success:
                etag = meta["etag"]
                if etag and method.upper() == "GET" and use_cache:
                    self.cache.set(cache_key, etag, body, dict(response.headers))
                return TransportResponse(
                    status_code=response.status_code,
                    data=body,
                    headers=dict(response.headers),
                    etag=etag,
                    from_cache=False,
                    request_id=meta["request_id"],
                    rate_limit=rate,
                    edition=meta["edition"],
                )

            raise error_from_response(
                status_code=response.status_code,
                url=str(response.url),
                body=body,
                request_id=meta["request_id"],
                retry_after=rate.retry_after,
            )

        # Should be unreachable, but guard against it.
        if last_exc is not None:
            raise FactorsAPIError(
                "Request failed after %d attempts: %s" % (self.max_retries, last_exc)
            ) from last_exc
        raise FactorsAPIError(
            "Request failed after %d attempts" % self.max_retries
        )

    def _sleep_backoff(self, attempt: int) -> None:
        time.sleep(self._compute_wait(attempt, None))

    @staticmethod
    def _compute_wait(attempt: int, retry_after: Optional[float]) -> float:
        if retry_after is not None and retry_after >= 0:
            return min(float(retry_after), 60.0)
        return min(2 ** (attempt - 1), 30.0)


# ---------------------------------------------------------------------------
# Async transport
# ---------------------------------------------------------------------------


class AsyncTransport:
    """Asynchronous HTTP transport backed by :class:`httpx.AsyncClient`."""

    def __init__(
        self,
        base_url: str,
        *,
        auth: Optional[AuthProvider] = None,
        timeout: float = DEFAULT_TIMEOUT,
        max_retries: int = DEFAULT_MAX_RETRIES,
        user_agent: str = DEFAULT_USER_AGENT,
        default_edition: Optional[str] = None,
        cache: Optional[ETagCache] = None,
        client: Optional[Any] = None,
        transport: Optional[Any] = None,
        extra_headers: Optional[Mapping[str, str]] = None,
    ) -> None:
        _check_httpx()
        self.base_url = base_url.rstrip("/")
        self.auth = auth
        self.timeout = timeout
        self.max_retries = max(1, int(max_retries))
        self.user_agent = user_agent
        self.default_edition = default_edition
        self.cache = cache if cache is not None else ETagCache()
        self._extra_headers = dict(extra_headers or {})
        self._owns_client = client is None
        if client is None:
            self._client = httpx.AsyncClient(
                base_url=self.base_url,
                timeout=timeout,
                transport=transport,
            )
        else:
            self._client = client

    async def aclose(self) -> None:
        if self._owns_client:
            await self._client.aclose()

    async def __aenter__(self) -> "AsyncTransport":
        return self

    async def __aexit__(self, *exc: Any) -> None:
        await self.aclose()

    def _headers(
        self,
        *,
        method: str,
        path: str,
        body: Optional[bytes],
        extra: Optional[Mapping[str, str]] = None,
    ) -> Dict[str, str]:
        base = {
            "Accept": "application/json",
            "User-Agent": self.user_agent,
        }
        base.update(self._extra_headers)
        if self.default_edition:
            base["X-Factors-Edition"] = self.default_edition
        if extra:
            base.update(extra)
        return compose_auth_headers(
            self.auth, base, method=method, path=path, body=body
        )

    async def request(
        self,
        method: str,
        path: str,
        *,
        params: Optional[Mapping[str, Any]] = None,
        json_body: Optional[Any] = None,
        extra_headers: Optional[Mapping[str, str]] = None,
        use_cache: bool = True,
    ) -> TransportResponse:
        import asyncio

        url_path = path if path.startswith("/") else "/" + path
        cache_key = ETagCache.key(method, url_path, params)

        body_bytes: Optional[bytes] = None
        if json_body is not None:
            body_bytes = json.dumps(json_body, separators=(",", ":"), sort_keys=True).encode(
                "utf-8"
            )

        cached = self.cache.get(cache_key) if (use_cache and method.upper() == "GET") else None
        cond_extra: Dict[str, str] = dict(extra_headers or {})
        if cached is not None and "If-None-Match" not in cond_extra:
            cond_extra["If-None-Match"] = cached[0]

        headers = self._headers(
            method=method, path=url_path, body=body_bytes, extra=cond_extra
        )
        if body_bytes is not None:
            headers["Content-Type"] = "application/json"

        last_exc: Optional[BaseException] = None
        for attempt in range(1, self.max_retries + 1):
            try:
                response = await self._client.request(
                    method,
                    url_path,
                    params=dict(params) if params else None,
                    content=body_bytes,
                    headers=headers,
                )
            except httpx.RequestError as exc:
                last_exc = exc
                logger.warning(
                    "Network error on %s %s (attempt %d/%d): %s",
                    method, url_path, attempt, self.max_retries, exc,
                )
                if attempt >= self.max_retries:
                    raise FactorsAPIError(
                        "Network error: %s" % exc,
                        context={"attempts": attempt, "url": url_path},
                    ) from exc
                await asyncio.sleep(Transport._compute_wait(attempt, None))
                continue

            meta = _extract_response_meta(response)
            rate: RateLimitInfo = meta["rate_limit"]

            if response.status_code == 304 and cached is not None:
                return TransportResponse(
                    status_code=200,
                    data=cached[1],
                    headers=dict(response.headers),
                    etag=cached[0],
                    from_cache=True,
                    request_id=meta["request_id"],
                    rate_limit=rate,
                    edition=meta["edition"],
                )

            if _should_retry(response.status_code) and attempt < self.max_retries:
                wait = Transport._compute_wait(attempt, rate.retry_after)
                logger.warning(
                    "Retrying %s %s after HTTP %d in %.2fs (attempt %d/%d)",
                    method, url_path, response.status_code, wait, attempt, self.max_retries,
                )
                await asyncio.sleep(wait)
                continue

            body = _decode_body(response)

            if response.is_success:
                etag = meta["etag"]
                if etag and method.upper() == "GET" and use_cache:
                    self.cache.set(cache_key, etag, body, dict(response.headers))
                return TransportResponse(
                    status_code=response.status_code,
                    data=body,
                    headers=dict(response.headers),
                    etag=etag,
                    from_cache=False,
                    request_id=meta["request_id"],
                    rate_limit=rate,
                    edition=meta["edition"],
                )

            raise error_from_response(
                status_code=response.status_code,
                url=str(response.url),
                body=body,
                request_id=meta["request_id"],
                retry_after=rate.retry_after,
            )

        if last_exc is not None:
            raise FactorsAPIError(
                "Request failed after %d attempts: %s" % (self.max_retries, last_exc)
            ) from last_exc
        raise FactorsAPIError(
            "Request failed after %d attempts" % self.max_retries
        )


__all__ = [
    "Transport",
    "AsyncTransport",
    "TransportResponse",
    "RateLimitInfo",
    "ETagCache",
    "DEFAULT_TIMEOUT",
    "DEFAULT_MAX_RETRIES",
    "DEFAULT_USER_AGENT",
]
